Improve handling of native dependencies

joshuagl opened the pull request:
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue**: #179 (partial fix, `securesystemslib.gpg` still needs to be addressed)

**Description of the changes being introduced by the pull request**:
* Makes colorama optional, add a warning to the log when it's not available but otherwise ssl functions as normal
* rename pyca_crypto_keys -> rsa_keys
* By default fallback to pure Python ed25519 certificate verification when the nacl module isn't available, rather than making the use of nacl an option at function call time.
* Ensures the public facing modules `interface` and `keys` are importable even when the native dependencies are not available
* Ensure that public functions using features of the native dependencies provide an actionable error message when they can't be used (due to the absence of native dependencies)

**Please verify and check that the pull request fulfils the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

joshuagl commented:
I need to figure out how to cleanly test `interface` and `keys` without native dependencies installed. 

Does it make sense to define a different test suite with a test for the public interfaces that verifies that all public functions either return successfully or return the expected `UnsupportedLibraryError`, we could then have a separate Tox environment without the native dependencies installed that runs this suite.

Is there a cleaner/simpler way to do it?

joshuagl commented:
I've got a partial implementation of a mechanism to test the public interfaces in 0ca75ef, does that look reasonable to you @lukpueh? If so I'll finish off that patch (it's currently incomplete and has some hard-coded paths).

lukpueh commented:
Thanks, @joshuagl! This looks very reasonable to me.

lukpueh commented:
So I just looked closer at your changes and they still seem very sensible. Thanks a lot, @joshuagl!

I realized that the name of pyca_crypto_keys.py is quite misleading. It suggests that it is securesystemslib's interface to pyca/cryptography, although other modules (ecdsa_keys.py, hash.py) also interface that library. Maybe it would be clearer and more accurate to call it `rsa_keys.py`, this would also make it consistent with the other low-level modules in securesystemslib, which both encapsulate key algorithms and not crypto libraries.

Another thing that I noticed is that pynacl import error handling is repeated on two levels of abstractions, i.e. in keys.py and in ed25519_keys.py. Do you think we could do it only once? Maybe we can draw the line that I described in https://github.com/secure-systems-lab/securesystemslib/issues/179#issuecomment-541024625 one level of abstraction further down. That is e.g.:


```python
# in ed25519_keys.py

NACL = True
try:
  import nacl

except ImportError:
  NACL = False

# ed25519_keys.sign() will raise an error if nacl does not exist
def sign():
  if not NACL:
    raise UnsupportedLibraryError()
  else: 
    nacl.ed25519_sign()

# ed25519_keys.verify() has a pure-python fallback. 
def verify():
  if not NACL:
    pure_python_ed25519_verify()
  else: 
    nacl.ed25519_sign()
```

This would make keys.py completely oblivious to the whole import story. And we could do the same thing for ecdsa_keys.py and rsa_keys.py (sic!), albeit without the pure-python fallback.

In your approach, OTOH, I see the advantage of failing early, i.e. raising an `UnsupportedLibraryError` before doing unnecessary work in any of the functions in keys.py.

I hope you don't mind my brainstorming here. I really appreciate your efforts and am curious about your thoughts. (cc @SantiagoTorres).

lukpueh commented:
Btw. since `securesystemslib.gpg` requires gpg and pyca/cryptography I also mentioned it in #179. But I suggest that we deal with it in a separate PR.

joshuagl commented:
Thanks for the constructive feedback @lukpueh 


> I realized that the name of pyca_crypto_keys.py is quite misleading. It suggests that it is securesystemslib's interface to pyca/cryptography, although other modules (ecdsa_keys.py, hash.py) also interface that library. 

Would you like to see that rename here, or in a separate PR?

> Another thing that I noticed is that pynacl import error handling is repeated on two levels of abstractions, i.e. in keys.py and in ed25519_keys.py. Do you think we could do it only once? Maybe we can draw the line that I described in [#179 (comment)](https://github.com/secure-systems-lab/securesystemslib/issues/179#issuecomment-541024625) one level of abstraction further down. That is e.g.:
> 
> ```python
> # in ed25519_keys.py
> 
> NACL = True
> try:
>   import nacl
> 
> except ImportError:
>   NACL = False
> 
> # ed25519_keys.sign() will raise an error if nacl does not exist
> def sign():
>   if not NACL:
>     raise UnsupportedLibraryError()
>   else: 
>     nacl.ed25519_sign()
> 
> # ed25519_keys.verify() has a pure-python fallback. 
> def verify():
>   if not NACL:
>     pure_python_ed25519_verify()
>   else: 
>     nacl.ed25519_sign()
> ```
> 
> This would make keys.py completely oblivious to the whole import story. And we could do the same thing for ecdsa_keys.py and rsa_keys.py (sic!), albeit without the pure-python fallback.
> 
> In your approach, OTOH, I see the advantage of failing early, i.e. raising an `UnsupportedLibraryError` before doing unnecessary work in any of the functions in keys.py.

I was torn on this. Doing the import handling a level lower has the benefits of:
* not handling the import at two levels
* keeping the code in the public interfaces simpler and moving the complexity (or at least ugly code) to the lower level modules

One downside to the approach is that the unit tests become more complex, as in many places we'll need to provide valid (enough) inputs to the higher-level functions so that the "unnecessary work" can be completed before the `UnsupportedLibraryError` is thrown by the lower level wrappers.

> I hope you don't mind my brainstorming here. I really appreciate your efforts and am curious about your thoughts. (cc @SantiagoTorres).

I don't mind at all. I very much appreciate the constructive discussion around these changes.
I think the more effort/cleaner high-level interfaces approach is the better one.

Assuming you agree, I'll revise this PR to:
* move the pynacl import error handling only to the lower-level modules
* rename pyca_crypto_keys.py -> rsa_keys.py

I'll also take a look at adding similar logic to `securesystemslib.gpg`, depending on how much work it looks to be I may handle it here or in a separate PR.

Does that seem like a reasonable approach?

joshuagl commented:
I didn't get around to addressing `securesystemslib.gpg` yet, but I think the rest of this PR is in reasonable shape for review.

joshuagl commented on .travis.yml:
```diff
@@ -6,10 +6,14 @@ matrix:
     - python: "3.6"
       env: TOXENV=py36
+    - python: "3.6"
+      env: TOXENV=purepy36
```
Agreed. I think I first made this change before I submitted #201 with 67174038465331b429a374390b55176324cb566c.

joshuagl commented on securesystemslib/ecdsa_keys.py:
```diff
@@ -121,6 +125,10 @@ def generate_public_and_private(scheme='ecdsa-sha2-nistp256'):
 
+  if not CRYPTO: # pragma: no cover
+    raise securesystemslib.exceptions.UnsupportedLibraryError(
+      'ECDSA key support requires the cryptography library')
```
The module-level constant is a good idea, I'll add that and ensure `UnsupportedLibraryError` is listed in the `<Exceptions>` section where appropriate. I'll also review line continuation throughout the PR.

joshuagl commented on securesystemslib/ecdsa_keys.py:
```diff
@@ -121,6 +125,10 @@ def generate_public_and_private(scheme='ecdsa-sha2-nistp256'):
     'securesystemslib.formats.PEMECDSA_SCHEMA', respectively.
   """
 
+  if not CRYPTO: # pragma: no cover
```
This is technically easily handled thanks to the versatility of coverage[[1]](https://coverage.readthedocs.io/en/coverage-5.0.3/config.html#config), but I held off on doing it because I wasn't sure how valuable coverage results would be for the purepy environments - at least in the current setup with only `check_public_interfaces.py` run in the purepy Tox environments we won't cover much of the code at all.

lukpueh replied:
I was thinking of a setup with mutual exclusive coverage exemption. Something like a `# pragma: purepy`, to mark lines/branches that are excluded from `py{27, 35, 36, 37, 38}`, but are also the only lines/branches considered for `purepy{27, 38}`, so that all lines must be tested by either one or the other. But let's do this in a separate PR (I'll add a note on #179).

joshuagl commented on tests/check_public_interfaces.py:
```diff
@@ -0,0 +1,203 @@
+
+    data = 'foo'
+    keydict = {'keytype': 'ed25519',
+                   'scheme': 'ed25519',
```
User error, I'll fix these and update.

joshuagl commented on securesystemslib/ed25519_keys.py:
```diff
@@ -155,6 +143,10 @@ def generate_public_and_private():
+  if not NACL: # pragma: no cover
+    raise securesystemslib.exceptions.UnsupportedLibraryError(
+      'ed25519 key support requires the nacl library')
+
```
Correct, I'll fix this.

joshuagl commented on securesystemslib/ed25519_keys.py:
```diff
@@ -237,6 +229,10 @@ def create_signature(public_key, private_key, data, scheme):
+  if not NACL: # pragma: no cover
+    raise securesystemslib.exceptions.UnsupportedLibraryError(
+      'ed25519 key support requires the nacl library')
+
```
Also correct.

joshuagl commented:
Thanks for the review @lukpueh. I think I've addressed all of your review comments except configuring coverage differently in the purepy case. I'd appreciate your thoughts on that one, I commented in-line.
